Speed up CUDA kernel launch when block/thread extents are statically known

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42901 Add a microbenchmark for LSTM elementwise portion
* #42900 Fix TE microbenchmark harness to use appropriate fuser/executor
* **#42899 Speed up CUDA kernel launch when block/thread extents are statically known**

Differential Revision: [D23078708](https://our.internmc.facebook.com/intern/diff/D23078708)

bertmaher committed:
Speed up CUDA kernel launch when block/thread extents are statically known

[ghstack-poisoned]

nickgg commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -855,15 +855,26 @@ void CudaCodeGen::call(const std::vector<CallArg>& args) {
   // TODO: eventually, codegen these calculations and make them part of the
   // module.
   for (size_t i = 0; i < gpu_block_extents.size(); i++) {
+    auto extent = dynamic_cast<const IntImm*>(gpu_block_extents[i]);
```
It's almost certainly an int, but supporting more dtypes doesn't cost us much here:

```
if (gpu_block_extents[i]->isConstant()) {
  gpu_block_extents_v[i] = immeditateAs<int>(gpu_block_extents[i]);
}
```

bertmaher committed:
Update on "Speed up CUDA kernel launch when block/thread extents are statically known"


Differential Revision: [D23078708](https://our.internmc.facebook.com/intern/diff/D23078708)

[ghstack-poisoned]

bertmaher committed:
Update on "Speed up CUDA kernel launch when block/thread extents are statically known"


Differential Revision: [D23078708](https://our.internmc.facebook.com/intern/diff/D23078708)

[ghstack-poisoned]
